File: DNA-Based Biomarker Discovery and Computational Drug Target Identification/scripts/dna_feature_extraction.py
"""
DNA Sequence Feature Extraction for Biomarker Discovery and Drug Target Identification

This script extracts comprehensive features from DNA sequences that can be used
to identify potential biomarkers and drug targets.

Author: DNA-Based Biomarker Discovery Project
Date: 2025
"""

import pandas as pd
import numpy as np
from typing import List, Tuple
import argparse
import sys
import os

# Add parent directory to path for imports
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))


def extract_dna_features(sequences: List[str]) -> np.ndarray:
    """
    Extract comprehensive features from DNA sequences.
    
    Features include:
    - Nucleotide composition (A, T, G, C frequencies)
    - GC content and GC skew
    - AT content and AT skew
    - Dinucleotide frequencies (16 combinations)
    - Trinucleotide frequencies (20 most common)
    - K-mer frequencies
    - Sequence complexity (Shannon entropy)
    - Purine/Pyrimidine ratio
    - Homopolymer runs
    - Sequence length (normalized)
    
    Parameters:
    -----------
    sequences : List[str]
        List of DNA sequences
        
    Returns:
    --------
    np.ndarray
        Feature matrix with shape (n_sequences, n_features)
    """
    features = []
    
    for seq in sequences:
        seq = seq.upper()
        seq_len = len(seq)
        
        if seq_len == 0:
            # Return zero vector for empty sequences
            features.append([0] * 48)
            continue
        
        # 1. Nucleotide composition
        a_count = seq.count('A') / seq_len
        t_count = seq.count('T') / seq_len
        g_count = seq.count('G') / seq_len
        c_count = seq.count('C') / seq_len
        
        # 2. GC content
        gc_content = (seq.count('G') + seq.count('C')) / seq_len
        
        # 3. AT content
        at_content = (seq.count('A') + seq.count('T')) / seq_len
        
        # 4. GC skew
        g_count_abs = seq.count('G')
        c_count_abs = seq.count('C')
        gc_skew = (g_count_abs - c_count_abs) / (g_count_abs + c_count_abs) if (g_count_abs + c_count_abs) > 0 else 0
        
        # 5. AT skew
        a_count_abs = seq.count('A')
        t_count_abs = seq.count('T')
        at_skew = (a_count_abs - t_count_abs) / (a_count_abs + t_count_abs) if (a_count_abs + t_count_abs) > 0 else 0
        
        # 6. Dinucleotide frequencies (16 possible combinations)
        dinucleotides = ['AA', 'AT', 'AG', 'AC', 'TA', 'TT', 'TG', 'TC', 
                        'GA', 'GT', 'GG', 'GC', 'CA', 'CT', 'CG', 'CC']
        dinuc_freq = [seq.count(dinuc) / (seq_len - 1) if seq_len > 1 else 0 for dinuc in dinucleotides]
        
        # 7. Trinucleotide frequencies (20 most common trinucleotides)
        trinucleotides = ['AAA', 'AAT', 'AAG', 'AAC', 'ATA', 'ATT', 'ATG', 'ATC',
                         'AGA', 'AGT', 'AGG', 'AGC', 'ACA', 'ACT', 'ACG', 'ACC',
                         'TAA', 'TAT', 'TAG', 'TAC']
        trinuc_freq = [seq.count(trinuc) / (seq_len - 2) if seq_len > 2 else 0 for trinuc in trinucleotides]
        
        # 8. Sequence complexity (Shannon entropy)
        nucleotide_counts = [seq.count('A'), seq.count('T'), seq.count('G'), seq.count('C')]
        nucleotide_probs = [count / seq_len for count in nucleotide_counts if seq_len > 0]
        shannon_entropy = -sum(p * np.log2(p) for p in nucleotide_probs if p > 0)
        
        # 9. Purine/Pyrimidine ratio
        purine_count = seq.count('A') + seq.count('G')
        pyrimidine_count = seq.count('C') + seq.count('T')
        pur_pyr_ratio = purine_count / pyrimidine_count if pyrimidine_count > 0 else 0
        
        # 10. Sequence length (normalized)
        normalized_length = seq_len / 100.0  # Assuming max length around 100
        
        # 11. Homopolymer runs (consecutive same nucleotides)
        max_homopolymer = max(
            len(max(seq.split('A'), key=len)) if 'A' in seq else 0,
            len(max(seq.split('T'), key=len)) if 'T' in seq else 0,
            len(max(seq.split('G'), key=len)) if 'G' in seq else 0,
            len(max(seq.split('C'), key=len)) if 'C' in seq else 0
        ) if seq_len > 0 else 0
        
        # Combine all features
        feature_vector = [
            a_count, t_count, g_count, c_count,
            gc_content, at_content, gc_skew, at_skew,
            shannon_entropy, pur_pyr_ratio, normalized_length, max_homopolymer
        ] + dinuc_freq + trinuc_freq
        
        features.append(feature_vector)
    
    return np.array(features)


def get_feature_names() -> List[str]:
    """
    Get names of all extracted features.
    
    Returns:
    --------
    List[str]
        List of feature names
    """
    feature_names = [
        'A_freq', 'T_freq', 'G_freq', 'C_freq',
        'GC_content', 'AT_content', 'GC_skew', 'AT_skew',
        'Shannon_entropy', 'Pur_Pyr_ratio', 'Normalized_length', 'Max_homopolymer'
    ]
    
    # Dinucleotide feature names
    dinucleotides = ['AA', 'AT', 'AG', 'AC', 'TA', 'TT', 'TG', 'TC', 
                    'GA', 'GT', 'GG', 'GC', 'CA', 'CT', 'CG', 'CC']
    feature_names.extend([f'Dinuc_{dinuc}' for dinuc in dinucleotides])
    
    # Trinucleotide feature names
    trinucleotides = ['AAA', 'AAT', 'AAG', 'AAC', 'ATA', 'ATT', 'ATG', 'ATC',
                     'AGA', 'AGT', 'AGG', 'AGC', 'ACA', 'ACT', 'ACG', 'ACC',
                     'TAA', 'TAT', 'TAG', 'TAC']
    feature_names.extend([f'Trinuc_{trinuc}' for trinuc in trinucleotides])
    
    return feature_names


def main():
    """Main function to extract features from DNA sequences."""
    parser = argparse.ArgumentParser(
        description='Extract features from DNA sequences for biomarker discovery'
    )
    parser.add_argument(
        '--input', 
        type=str, 
        default='../data/genomics_data.csv',
        help='Path to input CSV file with DNA sequences'
    )
    parser.add_argument(
        '--output', 
        type=str, 
        default='../results/dna_features.csv',
        help='Path to output CSV file with extracted features'
    )
    parser.add_argument(
        '--sequence-column',
        type=str,
        default='Sequences',
        help='Name of the column containing DNA sequences'
    )
    
    args = parser.parse_args()
    
    # Load data
    print(f"Loading data from {args.input}...")
    df = pd.read_csv(args.input)
    
    if args.sequence_column not in df.columns:
        raise ValueError(f"Column '{args.sequence_column}' not found in data. Available columns: {df.columns.tolist()}")
    
    # Extract features
    print("Extracting features from DNA sequences...")
    sequences = df[args.sequence_column].values.tolist()
    X = extract_dna_features(sequences)
    
    # Create feature DataFrame
    feature_names = get_feature_names()
    features_df = pd.DataFrame(X, columns=feature_names)
    
    # Add original columns if they exist (e.g., Labels)
    if 'Labels' in df.columns:
        features_df['Labels'] = df['Labels'].values
    
    # Save features
    print(f"Saving features to {args.output}...")
    os.makedirs(os.path.dirname(args.output), exist_ok=True)
    features_df.to_csv(args.output, index=False)
    
    print(f"Feature extraction complete!")
    print(f"  - Number of sequences: {len(sequences)}")
    print(f"  - Number of features: {X.shape[1]}")
    print(f"  - Features saved to: {args.output}")


if __name__ == "__main__":
    main()

File: DNA-Based Biomarker Discovery and Computational Drug Target Identification/scripts/dna_feature_extraction.R
# DNA Sequence Feature Extraction for Biomarker Discovery
# 
# This script extracts comprehensive features from DNA sequences that can be used
# to identify potential biomarkers and drug targets.
#
# Author: DNA-Based Biomarker Discovery Project
# Date: 2025

# Load required libraries
library(dplyr)
library(stringr)
library(data.table)

#' Extract comprehensive features from DNA sequences
#'
#' Features include:
#' - Nucleotide composition (A, T, G, C frequencies)
#' - GC content and GC skew
#' - AT content and AT skew
#' - Dinucleotide frequencies (16 combinations)
#' - Trinucleotide frequencies (20 most common)
#' - Sequence complexity (Shannon entropy)
#' - Purine/Pyrimidine ratio
#' - Homopolymer runs
#' - Sequence length (normalized)
#'
#' @param sequences Character vector of DNA sequences
#' @return Data frame with extracted features
extract_dna_features <- function(sequences) {
  n_seqs <- length(sequences)
  
  # Initialize feature vectors
  features <- data.frame(
    A_freq = numeric(n_seqs),
    T_freq = numeric(n_seqs),
    G_freq = numeric(n_seqs),
    C_freq = numeric(n_seqs),
    GC_content = numeric(n_seqs),
    AT_content = numeric(n_seqs),
    GC_skew = numeric(n_seqs),
    AT_skew = numeric(n_seqs),
    Shannon_entropy = numeric(n_seqs),
    Pur_Pyr_ratio = numeric(n_seqs),
    Normalized_length = numeric(n_seqs),
    Max_homopolymer = numeric(n_seqs)
  )
  
  # Dinucleotide frequencies
  dinucleotides <- c("AA", "AT", "AG", "AC", "TA", "TT", "TG", "TC",
                     "GA", "GT", "GG", "GC", "CA", "CT", "CG", "CC")
  for (dinuc in dinucleotides) {
    features[[paste0("Dinuc_", dinuc)]] <- numeric(n_seqs)
  }
  
  # Trinucleotide frequencies
  trinucleotides <- c("AAA", "AAT", "AAG", "AAC", "ATA", "ATT", "ATG", "ATC",
                      "AGA", "AGT", "AGG", "AGC", "ACA", "ACT", "ACG", "ACC",
                      "TAA", "TAT", "TAG", "TAC")
  for (trinuc in trinucleotides) {
    features[[paste0("Trinuc_", trinuc)]] <- numeric(n_seqs)
  }
  
  # Process each sequence
  for (i in seq_along(sequences)) {
    seq <- toupper(sequences[i])
    seq_len <- nchar(seq)
    
    if (seq_len == 0) {
      next
    }
    
    # Nucleotide composition
    features$A_freq[i] <- str_count(seq, "A") / seq_len
    features$T_freq[i] <- str_count(seq, "T") / seq_len
    features$G_freq[i] <- str_count(seq, "G") / seq_len
    features$C_freq[i] <- str_count(seq, "C") / seq_len
    
    # GC content
    features$GC_content[i] <- (str_count(seq, "G") + str_count(seq, "C")) / seq_len
    
    # AT content
    features$AT_content[i] <- (str_count(seq, "A") + str_count(seq, "T")) / seq_len
    
    # GC skew
    g_count <- str_count(seq, "G")
    c_count <- str_count(seq, "C")
    features$GC_skew[i] <- ifelse((g_count + c_count) > 0,
                                   (g_count - c_count) / (g_count + c_count),
                                   0)
    
    # AT skew
    a_count <- str_count(seq, "A")
    t_count <- str_count(seq, "T")
    features$AT_skew[i] <- ifelse((a_count + t_count) > 0,
                                   (a_count - t_count) / (a_count + t_count),
                                   0)
    
    # Dinucleotide frequencies
    for (dinuc in dinucleotides) {
      count <- str_count(seq, dinuc)
      features[[paste0("Dinuc_", dinuc)]][i] <- ifelse(seq_len > 1,
                                                        count / (seq_len - 1),
                                                        0)
    }
    
    # Trinucleotide frequencies
    for (trinuc in trinucleotides) {
      count <- str_count(seq, trinuc)
      features[[paste0("Trinuc_", trinuc)]][i] <- ifelse(seq_len > 2,
                                                          count / (seq_len - 2),
                                                          0)
    }
    
    # Shannon entropy
    nucleotide_counts <- c(
      str_count(seq, "A"),
      str_count(seq, "T"),
      str_count(seq, "G"),
      str_count(seq, "C")
    )
    nucleotide_probs <- nucleotide_counts / seq_len
    nucleotide_probs <- nucleotide_probs[nucleotide_probs > 0]
    features$Shannon_entropy[i] <- -sum(nucleotide_probs * log2(nucleotide_probs))
    
    # Purine/Pyrimidine ratio
    purine_count <- str_count(seq, "A") + str_count(seq, "G")
    pyrimidine_count <- str_count(seq, "C") + str_count(seq, "T")
    features$Pur_Pyr_ratio[i] <- ifelse(pyrimidine_count > 0,
                                         purine_count / pyrimidine_count,
                                         0)
    
    # Normalized length
    features$Normalized_length[i] <- seq_len / 100.0
    
    # Max homopolymer run
    max_homopolymer <- 0
    for (nuc in c("A", "T", "G", "C")) {
      if (grepl(nuc, seq)) {
        parts <- strsplit(seq, paste0("[^", nuc, "]+"))[[1]]
        parts <- parts[parts != ""]
        if (length(parts) > 0) {
          max_run <- max(nchar(parts))
          max_homopolymer <- max(max_homopolymer, max_run)
        }
      }
    }
    features$Max_homopolymer[i] <- max_homopolymer
  }
  
  return(features)
}

#' Main function to extract features from DNA sequences
main <- function() {
  # Parse command line arguments
  args <- commandArgs(trailingOnly = TRUE)
  
  if (length(args) < 1) {
    input_file <- "../data/genomics_data.csv"
  } else {
    input_file <- args[1]
  }
  
  if (length(args) < 2) {
    output_file <- "../results/dna_features.csv"
  } else {
    output_file <- args[2]
  }
  
  # Load data
  cat("Loading data from", input_file, "...\n")
  df <- read.csv(input_file, stringsAsFactors = FALSE)
  
  if (!"Sequences" %in% colnames(df)) {
    stop("Column 'Sequences' not found in data. Available columns: ", 
         paste(colnames(df), collapse = ", "))
  }
  
  # Extract features
  cat("Extracting features from DNA sequences...\n")
  sequences <- df$Sequences
  features <- extract_dna_features(sequences)
  
  # Add labels if they exist
  if ("Labels" %in% colnames(df)) {
    features$Labels <- df$Labels
  }
  
  # Create output directory if it doesn't exist
  output_dir <- dirname(output_file)
  if (!dir.exists(output_dir)) {
    dir.create(output_dir, recursive = TRUE)
  }
  
  # Save features
  cat("Saving features to", output_file, "...\n")
  write.csv(features, output_file, row.names = FALSE)
  
  cat("Feature extraction complete!\n")
  cat("  - Number of sequences:", length(sequences), "\n")
  cat("  - Number of features:", ncol(features), "\n")
  cat("  - Features saved to:", output_file, "\n")
}

# Run main function if script is executed directly
if (!interactive()) {
  main()
}

